mypy/newsemanal/semanal.py

```diff
@@ -4052,15 +4063,19 @@ def add_symbol_table_node(self, name: str, symbol: SymbolTableNode,
         existing = names.get(name)
         if (existing is not None
                 and context is not None
-                and not isinstance(existing.node, PlaceholderNode)):
-            if existing.node != symbol.node:
+                and (not isinstance(existing.node, PlaceholderNode)
+                     or isinstance(symbol.node, PlaceholderNode))):
+            if (existing.node != symbol.node
+                    and not (isinstance(existing.node, PlaceholderNode)
+                             and isinstance(symbol.node, PlaceholderNode))):
```

Comment:
TBH, this bunch of `if`s is hard to read, I think it makes sense to either refactor it, or add explanatory comments.